posts/2019-10-28-nll-hard-errors.md

```diff
@@ -0,0 +1,107 @@
+---
+layout: post
+title: "Completing the transition to the new borrow checker"
+author: Niko Matsakis
+---
+
+For most of 2018, we've been issuing warnings about various bugs in the
+borrow checker that we plan to fix -- about two months ago, in the
+current Rust nightly, those warnings became **hard errors**. In about
+two weeks, when the nightly branches to become beta, those hard errors
+will be in the beta build, and they will eventually hit stable on
+December 19th, as part of Rust 1.40.0. **If you're testing with
+Nightly, you should be all set -- but otherwise, you may want to go
+and check to make sure your code still builds. If not, we have advice
+for fixing common problems below.**
+
+### Background: the non-lexical lifetime transition
+
+When we [released Rust 2018 in Rust 1.31][2018], it included a new
+version of the borrow checker, one that implemented ["non-lexical
+lifetimes"][nll]. This new borrow checker did a much more precise
+analysis than the original, allowing us to eliminate a lot of
+unnecessary errors and make Rust easier to use. I think most everyone
+who was using Rust 2015 can attest that this shift was a big
+improvement.
+
+### The new borrow checker also fixed a lot of bugs
+
+What is perhaps less well understood is that the new borrow checker
+implementation *also* fixed a lot of bugs. In other words, the new
+borrow checker did not just accept more programs -- **it also rejected
+some programs that never should have been accepted in the first
+place!**
+
+[2018]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html
+[nll]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#non-lexical-lifetimes
+[MIR]: https://blog.rust-lang.org/2016/04/19/MIR.html
+
+### Until recently, those fixed bugs produced warnings, not errors
+
+As part of our commitment to stability, whenever we find bugs that
+impact existing code in a major way, we try to "phase in" those
+changes gradually. We usually begin with "Future Compatibility
+Warnings", for example, before moving those warnings to hard errors
+(sometimes a small bit at a time). Since the bug fixes to the borrow
+checker affected a lot of crates, we knew we needed a warning period
+before we could make them into hard errors.
+
+To implement this warning period, we kept two copies of the borrow
+checker around (this is a trick we use quite frequently, actually).
+The new checker ran first. If it found errors, we didn't report them
+directly: instead, we ran the old checker in order to see if the crate
```

Comment:
```suggestion
directly: instead, we ran the old checker to see if the crate
```

Reply:
I'm going to push back here. I realize "in order" is not necessary, but I find it works in the flow of this sentence for reasons I find hard to *precisely* articulate. Somehow "ran the old checker to see if the crate used to compile" feels like it "packs a lot in", requiring more mental effort for me to parse and understand, whereas "in order to see" clarifies the overall structure of the sentence and is easier.

Reply:
I think this is fine as-is, and generally prefer the author's preference when it comes to stuff like this.

Reply:
It's your blog post, but personally I find wordiness bad in technical writing. See also:

- https://www.oxford-royale.com/articles/clumsy-phrases-avoid.html
- https://blog.apastyle.org/apastyle/2015/09/principles-of-writing-how-to-avoid-wordiness.html
- https://proofreadingpal.com/proofreading-pulse/essays/how-to-avoid-wordiness/
- https://www.bcit.ca/files/ltc/pdf/avoid_in_order_to.pdf